File: hex_terrain_editor/model/path_feature.py
"""Path features: roads, rivers, streams as polyline/Bézier features."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any

from .terrain_types import PathType


@dataclass
class PathFeature:
    """A polyline feature (road, river, etc.) that crosses hex boundaries."""

    id: str = field(default_factory=lambda: str(uuid.uuid4())[:8])
    feature_type: PathType = PathType.ROAD
    control_points: list[tuple[float, float]] = field(default_factory=list)
    width: float = 2.0
    properties: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "type": self.feature_type.value,
            "control_points": [list(p) for p in self.control_points],
            "width": self.width,
            "properties": self.properties,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> PathFeature:
        return cls(
            id=data["id"],
            feature_type=PathType(data["type"]),
            control_points=[tuple(p) for p in data["control_points"]],
            width=data.get("width", 2.0),
            properties=data.get("properties", {}),
        )

    def bounding_box(self) -> tuple[float, float, float, float]:
        """Return (min_x, min_y, max_x, max_y) of the control points."""
        if not self.control_points:
            return (0, 0, 0, 0)
        xs = [p[0] for p in self.control_points]
        ys = [p[1] for p in self.control_points]
        margin = self.width * 2
        return (min(xs) - margin, min(ys) - margin, max(xs) + margin, max(ys) + margin)

File: src/model/path_feature.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use super::terrain_types::PathType;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PathFeature {
    pub id: String,
    pub feature_type: PathType,
    pub control_points: Vec<[f32; 2]>,
    pub width: f32,
    #[serde(default)]
    pub properties: serde_json::Value,
}

impl PathFeature {
    pub fn new(feature_type: PathType) -> Self {
        let id = Uuid::new_v4().to_string();
        let short_id = id[..8].to_string();
        Self {
            id: short_id,
            feature_type,
            control_points: Vec::new(),
            width: feature_type.width(),
            properties: serde_json::Value::Null,
        }
    }

    /// Compute axis-aligned bounding box of the path, padded by `width * 2`.
    /// Returns `(min_x, min_y, max_x, max_y)`.
    pub fn bounding_box(&self) -> (f32, f32, f32, f32) {
        if self.control_points.is_empty() {
            return (0.0, 0.0, 0.0, 0.0);
        }
        let mut min_x = f32::MAX;
        let mut min_y = f32::MAX;
        let mut max_x = f32::MIN;
        let mut max_y = f32::MIN;
        for pt in &self.control_points {
            min_x = min_x.min(pt[0]);
            min_y = min_y.min(pt[1]);
            max_x = max_x.max(pt[0]);
            max_y = max_y.max(pt[1]);
        }
        let pad = self.width * 2.0;
        (min_x - pad, min_y - pad, max_x + pad, max_y + pad)
    }
}
